File: modules/rtm/R/truncnorm.R
#' Truncated normal distribution functions
#' 
#' @name rtnorm
#' @title Random sampling from one-sided truncated normal distribution
#' @author Alexey Shiklomanov
#' @details Draws a random number and, if it doesn't fall within the 
#'      specified range, resample using an adjusted Normal CDF. This isn't
#'      performed immediately because CDF sampling calls three functions --
#'      qnorm, runif, and pnorm--and therefore is much less efficient than a
#'      simple random sample.
#' @param mu The mean parameter of the truncated normal. NOTE that this is NOT
#'      equal to the mean of the distribution
#' @param sd The standard deviation parameter of the truncated normal. Again, NOTE
#'      tht this is not the SD of the distribution.
#' @param MIN The minimum value, which defines the truncation.
#' @return Numeric length one.
rtnorm <- function(mu, sd, MIN){
    x <- rnorm(1, mu, sd)
    if(x < MIN)
        x <- qnorm(runif(1, pnorm(MIN, mu, sd, 1, 0), 1), mu, sd, 1, 0)
    return(x)
}

#' @name dtnorm
#' @title Truncated normal distribution density
#' @author Alexey Shiklomanov
#' @details Calculates the log density of a univariate truncated normal variable
#' @param x A random variable
#' @param mu The mean parameter of the distribution; NOTE this is not equal to the mean
#' @param sd The standard deviation parameter of the distribution
#' @param MIN Value at which the truncation takes place
#' @return Numeric; log density of the distribution, or -1e15 if the x < MIN
dtnorm <- function(x, mu, sd, MIN){
    if(x < MIN)
        return(-1e15)
    else
        return(dnorm(x, mu, sd, 1) - log(1-pnorm(MIN, mu, sd, 1, 0)))
}

File: modules/rtm/src/truncnorm.cpp
#include <Rcpp.h>
#include <Rmath.h>
using namespace Rcpp;

double rtnorm(double mu, double sd, double MIN){
    double x = rnorm(1, mu, sd)[0];
    if(x < MIN){
        x = R::qnorm(runif(1, R::pnorm(MIN, mu, sd, 1, 0), 1)[0], mu, sd, 1, 0);
    }
    return x;
}

double dtnorm(double X, double mu, double sd, double MIN){
    if(X < MIN){
        return -1e15;
    }
    else {
        return R::dnorm(X, mu, sd, 1) - log(1 - R::pnorm(MIN, mu, sd, 1, 0));
    }
}

double rtnorm_c(double mu, double sd, double MIN){
    double M, S, L_min;
    L_min = (R::dnorm(MIN, mu, sd, 0))/(1 - R::pnorm(MIN, mu, sd, 1, 0));
    M = mu - sd*L_min;
    S = sqrt(pow(sd,2) / (1 - L_min * (L_min - (MIN - mu)/sd)));
    return rtnorm(M, S, MIN);
}

double dtnorm_c(double X, double mu, double sd, double MIN){
	if(X < MIN){
		return -1e15;
	}
	else {	
		double M, S, L_min;
		L_min = (R::dnorm(MIN, mu, sd, 0))/(1 - R::pnorm(MIN, mu, sd, 1, 0));
		M = mu - sd*L_min;
		S = sqrt(pow(sd,2) / (1 - L_min * (L_min - (MIN - mu)/sd)));
		return dtnorm(X, M, S, MIN);
	}
}
